src/net/server/handlers/create_query.rs

```diff
@@ -0,0 +1,246 @@
+};
+use async_trait::async_trait;
+use axum::{
+    body::{Bytes, HttpBody},
+    extract::{BodyStream, FromRequest, Query, RequestParts},
+};
+use futures::{ready, Stream};
+use pin_project::pin_project;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// the string repr of the field type from the query param of the request
+/// Possible values are currently:
+/// * `fp2`
+/// * `fp31`
+/// * `fp32_bit_prime`
+#[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+struct FieldStr {
+    field_type: String,
+}
+
+/// Wraps a [`BodyStream`] and produce a new stream that has chunks of exactly size `size_in_bytes`.
+/// # Errors
+/// If the downstream body is not a multiple of `size_in_bytes`.
+#[pin_project]
+pub struct ByteArrStream {
+    #[pin]
+    body: BodyStream,
+    size_in_bytes: u32,
+    buffered_size: u32,
+    buffered: Vec<Bytes>,
+}
+
+impl ByteArrStream {
+    pub fn new(body: BodyStream, size_in_bytes: u32) -> Self {
+        Self {
+            body,
+            size_in_bytes,
+            buffered_size: 0,
+            buffered: Vec::new(),
+        }
+    }
+
+    /// returns [`Bytes`] of `size_in_bytes` length from the buffered chunks
+    /// # Panics
+    /// if the total length of byte buffers is smaller than `size_in_bytes`
+    fn take_bytes(size_in_bytes: u32, buffered: &mut Vec<Bytes>, buffered_size: &mut u32) -> Bytes {
+        assert!(*buffered_size >= size_in_bytes);
+
+        // take from single buffer
+        // simply splits the first buffer in-place
+        if buffered[0].len() > size_in_bytes as usize {
+            let out_bytes = buffered[0].split_to(size_in_bytes as usize);
+            *buffered_size -= size_in_bytes;
+            out_bytes
+        } else {
+            // take across buffers
+            // removes expended buffers after taking
+            let mut out_bytes = Vec::new();
+            let mut remaining_bytes = size_in_bytes as usize;
+            let mut i = 0;
+            loop {
+                if remaining_bytes == 0 {
+                    break;
+                }
+                let buffer = &mut buffered[i];
+                let buffer_len = buffer.len();
+                // current `buffer` has more bytes than needed
+                if buffer_len > remaining_bytes {
+                    out_bytes.push(buffer.split_to(remaining_bytes));
+                    remaining_bytes = 0;
+                } else {
+                    // current `buffer` has <= bytes needed, needs removal
+                    out_bytes.push(buffer.clone());
+                    i += 1;
+                    remaining_bytes -= buffer.len();
+                }
+            }
+
+            // remove expended buffers
+            buffered.drain(0..i);
+            // reduce size of total buffers accordingly
+            *buffered_size -= size_in_bytes;
+            // create new `Bytes` that combines chunks
+            let bytes_vec = out_bytes.into_iter().fold(
```

Comment:
I see you worked hard to keep it zero-copy (and I applaud the efforts!) but this is where all your efforts are nullified as you unconditionally allocate here. I believe your approach results in 2 memcopies (one in `out_bytes` and one here). Note that you could make it almost always zero copy by:

* checking if buffer has enough capacity for the next slice. If it does, you simply do `split_to` and **replace** the buffer with the remaining portion. Now you can return the result of `split_to` and you don't need to copy anything. Cost - O(1)
* if you don't have enough capacity in your buffer, then you need to allocate, but only for this operation. Consume the current buffer and `split_to` the next one to create a contiguous slice. Cost - memalloc but the next few operations the cost will be O(1) again. 

There is a slightly more expensive approach that always does 1 memcopy but easier to implement. You could also keep a large buffer of `BytesMut` and do a memcopy when you consume the next chunk from the upstream. That costs you something, but pulling next chunk is a one-liner `split_to` followed by `frozen`. 

I think you're pretty close to make things as optimal as they could be, but wanted to show you an easier approach that may be acceptable to start with.

Reply:
* `out_bytes` is a vec of `Bytes`, so it's only allocating for the pointers to the `Bytes`; the reason it's a vec is due to the possibility of very small `Bytes` from the body, however unlikely. e.g.:
  * body looks like: `[Bytes[1], Bytes[1], Bytes[2]]` (2 `Bytes` of length 1 u8, 1 `Bytes` of length 2 u8)
  * `size_in_bytes` is 4
  * I need all 3 `Bytes` to fulfill `take_bytes`
  * in this case, there's still only 1 memcopy of u8's per `Bytes` array (inside the `fold`)
*  the first thing this function does is check if the first `Bytes` is larger than the size, and if so, does exactly as you say for O(1) time. only when the first buffer is smaller does it resort to the memcopy
* > Consume the current buffer and split_to the next one to create a contiguous slice
  
  this is the current behavior; tho to be clear, it memcopies both bufs into a new contiguous one

I'll see if i can improve the comments to make it more clear what's happening

Reply:
> out_bytes is a vec of Bytes, so it's only allocating for the pointers to the Bytes;

true, however Bytes take 32 bytes in memory, while field pair is 8 in the worst case of 32 bit prime, so it is even more expensive

but I think you fixed that part now and now it is much cleaner

Reply:
yeah, that's a good point. writing directly to byte vec is much better, so thanks for the tip

Reply:
we should keep that in mind when polling from this stream. Getting 32 bytes pointer to 8 bytes is suboptimal, we should have chunks of a much larger size